src/game/collision.cpp

```diff
@@ -129,10 +231,60 @@ void CCollision::MovePoint(vec2 *pInoutPos, vec2 *pInoutVel, float Elasticity, i
 		*pInoutPos = Pos + Vel;
 	}
 }
+void CCollision::Diffract(vec2* pInoutPos, vec2* pInoutVel, float Elasticity, int* pBounces, int Flag) const
+{
+	if (pBounces)
+		*pBounces = 0;
+
+	vec2 Pos = *pInoutPos;
+	vec2 Vel = *pInoutVel;
+	if (CheckPoint(Pos + Vel, Flag))
+	{
+		int Affected = 0;
+		if (CheckPoint(Pos.x + Vel.x, Pos.y, Flag))
+		{
+			pInoutVel->x *= -Elasticity;
+			pInoutPos->x += clamp(-Elasticity * Vel.x, -1.0f, 1.0f);
+			if (pBounces)
+				(*pBounces)++;
+			Affected++;
+		}
+
+		if (CheckPoint(Pos.x, Pos.y + Vel.y, Flag))
+		{
+			pInoutVel->y *= -Elasticity;
+			pInoutPos->y += clamp(-Elasticity * Vel.y, -1.0f, 1.0f);
+			if (pBounces)
+				(*pBounces)++;
+			Affected++;
+		}
+
```

Comment:
@Stiopa866 what happens if I shoot a laser perfectly at a corner? 🤔 never tried that

Reply:
In general or when shooting into water? :P